test/distributed/test_c10d_common.py

```diff
@@ -1326,6 +1336,100 @@ def test_send_recv(self):
 instantiate_parametrized_tests(CommonDistributedDataParallelTest)
 
 
+def wait_comm(comm_tensor):
+    comm_tensor._work.wait()
+    return comm_tensor._tensor
+
+
+class CommTensor(torch.Tensor):
+    @staticmethod
+    def __new__(cls, tensor: torch.Tensor, work: torch.distributed._Work):
+
+        r = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
+            cls,
+            tensor.size(),
+            dtype=tensor.dtype,
+            device=tensor.device,
+            layout=tensor.layout,
+            requires_grad=tensor.requires_grad,
+        )
+        r._tensor = tensor
+        r._work = work
+        return r
+
+    def __repr__(self):
+        return f"CommTensor({self._tensor})"
+
+    # disable __torch_function__ so that CommTensor can recursively dispatch
+    # ProxyTensor in make_fx
+    __torch_function__ = _disabled_torch_function_impl
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+
+        def unwrap(e):
+            if isinstance(e, CommTensor):
+                t = e._tensor
+                w = e._work
+
+                if isinstance(t, ProxyTensor):
+                    # in tracing mode, add wait_comm node to graph
+                    proxy_res = t.proxy_mode.tracer.create_proxy(
+                        'call_function',
+                        wait_comm,
+                        (e,),
+                        {},
+                        name="wait_comm"
+                    )
+                    t.proxy = proxy_res
+                    return t
+                else:
+                    # in eager mode, simply wait
+                    w.wait()
+                    return t
+            else:
+                return e
+
+        args = tree_map(unwrap, args)
+        kwargs = tree_map(unwrap, kwargs)
+
+        return func(*args, **kwargs)
+
+
+class CompilerTest(MultiProcessTestCase):
+    def setUp(self):
+        super(CompilerTest, self).setUp()
+        self._spawn_processes()
+
+    def tearDown(self):
+        super(CompilerTest, self).tearDown()
+        try:
+            os.remove(self.file_name)
+        except OSError:
+            pass
+
+    def _get_process_group(self):
+        raise NotImplementedError("To be implemented by subclass")
+
+    def _test_work_wait(self, x):
+        pg = self._get_default_group()
+
+        def fn(x):
+            y = x + x
+            work = dist.all_reduce(y, group=pg, async_op=True)
+            y = CommTensor(y, work)
+            return y * 2
+
+        xx = x.clone()
+
+        fx_fn = make_fx(fn)(xx)
+
+        y = fn(x)
+        yy = fx_fn(xx).elem
```

Comment:
@Chillee is there a better way to run the traced `fn` and extract the result?

Reply:
This doesn’t seem right… the traced graph shouldn’t have any notion of proxytensor.

Reply:
hmm, `fx_fn(xx)` does return a `ProxyTensor`.

Reply:
oh, it's because I called `t.proxy_mode.tracer.create_proxy` during tracing

Reply:
it memorized the ProxyTensor during tracing, and use that same tensor instance during execution, instead of correcting the new node to the output from the previous node. Digging